### PR TITLE
deprecate Dates.recur in favour of filter

### LINKED ISSUE #17513: `recur` fails when the `DateFunction` provided returns false outside of the range
The way `recur` is currently written requires that it be able to find at least one case where the `DateFunction` returns `true` once it steps outside of the `StepRange` provided. For this reason, even though these two examples should produce identical results, the second one throws an error:

``` julia
julia> Dates.recur(Date(1980):Dates.Day(-1):Date(1970)) do d
           Dates.ismonday(d) && d >= Date(1969)
       end
522-element Array{Date,1}:
 1979-12-31
 1979-12-24
 1979-12-17
 ⋮         
 1970-01-19
 1970-01-12
 1970-01-05

julia> Dates.recur(Date(1980):Dates.Day(-1):Date(1970)) do d
           Dates.ismonday(d) && d >= Date(1970)
       end
ERROR: ArgumentError: Adjustment limit reached: 10000 iterations
 in adjust(::Base.Dates.DateFunction, ::Date, ::Base.Dates.Day, ::Int64) at ./dates/adjusters.jl:148
 in #recur#24(::Base.Dates.Day, ::Bool, ::Int64, ::Function, ::Function, ::Date, ::Date) at ./dates/adjusters.jl:282
 in (::Base.Dates.#kw##recur)(::Array{Any,1}, ::Base.Dates.#recur, ::Function, ::Date, ::Date) at ./<missing>:0
 in recur(::Function, ::StepRange{Date,Base.Dates.Day}) at ./dates/adjusters.jl:299
 in eval(::Module, ::Any) at ./boot.jl:234
 in macro expansion at ./REPL.jl:92 [inlined]
 in (::Base.REPL.##1#2{Base.REPL.REPLBackend})() at ./event.jl:46
```

This happens because `adjust` is called again even when `cmp(next, stop)` returns `0`. If that call fails (because the `DateFunction` doesn't return true within `limit` attempts) then an error results, even though we're already done filling our array of results.

Of course, here the bug could be avoided by simply removing the second check from the `DateFunction`. The case in which I encountered this problem is a little more complex, but I think the point is adequately demonstrated by this example.

I'm running a random recent build (`0.5.0-dev+5275`; sorry!) but I've verified that this is present on `0.4.5` as well. Here's my current version information:

``` julia
julia> versioninfo()
Julia Version 0.5.0-dev+5275
Commit 45400d0 (2016-07-11 13:46 UTC)
Platform Info:
  System: Darwin (x86_64-apple-darwin15.5.0)
  CPU: Intel(R) Core(TM) i7-4558U CPU @ 2.80GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.7.1 (ORCJIT, haswell)
```

### PR DESCRIPTION
**Pull request by simonbyrne**
Should also indirectly close #17513

cc: @quinnj

### CODE REVIEW

**Comment by quinnj** [comment]
+1

File: base/deprecated.jl [hunk 1]
```diff
@@ -1098,4 +1098,16 @@ eval(Base.LinAlg, quote
     end
 end)
 
+eval(Base.Dates, quote
+    function recur{T<:TimeType}(fun::Function, dr::StepRange{T};negate::Bool=false, limit::Int=10000)
+        depwarn(string("Dates.recur is deprecated, use filter instead."),:recur)
```
Comment (ararslan) [inline base/deprecated.jl]: The `string()` part shouldn't be necessary, right?

Comment (simonbyrne) [inline base/deprecated.jl]: Good point, fixed.

File: base/deprecated.jl [hunk 2]
```diff
@@ -1102,4 +1102,17 @@ end)
 @deprecate den denominator
 @deprecate num numerator
 
+# #19288
+eval(Base.Dates, quote
+    function recur{T<:TimeType}(fun::Function, dr::StepRange{T};negate::Bool=false, limit::Int=10000)
```
Comment (StefanKarpinski) [inline base/deprecated.jl]: `;negate` => `; negate`

**Review by StefanKarpinski — APPROVED** [review]
Yes, much better for this to be part of `filter`.

**Comment by omus** [comment]
The manual will need to be updated. Particularly [doc/manual/dates.rst](http://julia.readthedocs.io/en/latest/manual/dates/#adjuster-functions)

**Comment by simonbyrne** [comment]
Good point. Is it worth keeping the text, or should we just get rid of it altogether?

**Comment by quinnj** [comment]
I think it's worth mentioning, particularly since there used to be explicit functionality for this.

**Comment by tkelman** [comment]
There was a docstring for this that genstdlib is now warning about. Should that be repurposed as a dates-specific docstring for filter?

**Comment by tkelman** [comment]
Bump @simonbyrne there's a stale docstring here giving a warning, should be cleaned up especially if we're about to transition how the docs get built.

**Comment by simonbyrne** [comment]
No, I don't think we need anything date-specific. I'll open a PR to fix it.

**Pull request merged**